utils/result_try.hh

```diff
@@ -0,0 +1,442 @@
+    }
+};
+
+static_assert(ExceptionHandle<failed_result_handle<dummy_result<>>, dummy_result<>>);
+
+template<ExceptionContainerResult R>
+struct exception_ptr_handle {
+private:
+    std::exception_ptr _eptr;
+
+public:
+    exception_ptr_handle(std::exception_ptr&& eptr) : _eptr(std::move(eptr)) {}
+
+    // Throws the exception.
+    [[noreturn]] R into_result() {
+        std::rethrow_exception(std::move(_eptr));
+    }
+
+    seastar::future<R> into_future() {
+        return seastar::make_exception_future<R>(std::move(_eptr));
+    }
+};
+
+// Check that it is a valid handle with some arbitrary ExceptionContainerResult
+static_assert(ExceptionHandle<exception_ptr_handle<dummy_result<>>, dummy_result<>>);
+
+template<typename Ex, typename Cb>
+struct result_catcher {
+private:
+    Cb _cb;
+
+public:
+    result_catcher(Cb&& cb) : _cb(std::move(cb)) {}
+
+    template<ExceptionContainerResult R, typename Converter, typename Continuation>
+    auto handle_exception_from_result(const auto& ex, R& original_result, Continuation&& cont)
+            -> typename Converter::template wrapped_type<R> {
+        if constexpr (std::is_base_of_v<Ex, std::remove_cvref_t<decltype(ex)>>) {
+            if constexpr (std::is_invocable_v<Cb, const Ex&, failed_result_handle<R>>) {
+                // Invoke with exception reference and handle
+                return Converter::template invoke(_cb, ex, failed_result_handle<R>(std::move(original_result)));
+            } else {
+                // Simplified interface - invoke with reference to exception only
+                static_assert(std::is_invocable_v<Cb, const Ex&>,
+                        "The handler function does not have a suitable call operator");
+                return Converter::template invoke(_cb, ex);
+            }
+        } else {
+            // Let another handler try it
+            return cont();
+        }
+    }
+
+    template<ExceptionContainerResult R, typename Converter, typename Continuation>
+    auto wrap_in_catch(Continuation&& cont) {
+        return [this, cont = std::move(cont)] () mutable -> typename Converter::template wrapped_type<R> {
+            try {
+                return cont();
+            } catch (const Ex& ex) {
+                if constexpr (std::is_invocable_v<Cb, const Ex&, exception_ptr_handle<R>>) {
+                    // Invoke with exception reference and handle
+                    return Converter::template invoke(_cb, ex, exception_ptr_handle<R>(std::current_exception()));
+                } else {
+                    // Simplified interface - invoke with reference to exception only
+                    static_assert(std::is_invocable_v<Cb, const Ex&>,
+                            "The handler function does not have a suitable call operator");
+                    return Converter::template invoke(_cb, ex);
+                }
+            }
+        };
+    }
+};
+
+template<typename Cb>
+struct result_catcher_dots {
```

Comment:
We would like to avoid converting failed results to exceptions. If you need to rethrow the exception, you should return `into_result` or `into_future` instead.

One thing that I noticed which is missing but will be useful to me when converting the read path is the ability to pass the the exception/failed result to a promise. I'll add it.

Are there some other uses of `std::exception_ptr` you can see but are not handled by this set of methods?